fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingThrottle.java

```diff
@@ -72,95 +71,186 @@ public class IndexingThrottle {
-        final int increaseLimitAfter = common.config.getIncreaseLimitAfter();
-        final int maxLimit = common.config.getMaxLimit();
-        if (increaseLimitAfter > 0) {
+        void handleLimitsPostRunnerTransaction(@Nullable Throwable exception,
+                                               @Nonnull final AtomicLong recordsScanned,
+                                               final boolean duringRangesIteration) {
+            final long recordsScannedThisTransaction = recordsScanned.get();
+            if (!duringRangesIteration) {
+                if (exception == null) {
+                    synchronized (this) { // In this mode, multi threads are allowed
+                        totalRecordsScannedSuccess += recordsScannedThisTransaction;
+                    }
+                }
+                return; // no adjustments here
+            }
+            // Here: assuming a single thread
             if (exception == null) {
-                successCount++;
-                if (successCount >= increaseLimitAfter && limit < maxLimit) {
+                countSuccessfulTransactions++;
+                totalRecordsScannedSuccess += recordsScannedThisTransaction;
+                recordsScannedSinceForcedDelay += recordsScannedThisTransaction;
+                if (consecutiveSuccessCount >= common.config.getIncreaseLimitAfter()) {
                     increaseLimit();
+                    consecutiveSuccessCount = 0; // do not increase again immediately after the next success
+                } else {
+                    consecutiveSuccessCount++;
                 }
             } else {
-                successCount = 0;
+                // Here: memorize the actual records count for the decrease limit function (if applicable) and reset the counter
+                countRunnerFailedTransactions++;
+                lastFailureRecordsScanned = recordsScannedThisTransaction;
+                totalRecordsScannedFailure += recordsScannedThisTransaction;
+                recordsScanned.set(0);
+            }
+        }
+
+        private void increaseLimit() {
+            final long maxLimit = common.config.getMaxLimit();
+            if (recordsLimit >= maxLimit) {
+                return; // quietly
+            }
+            final long oldLimit = recordsLimit;
+            recordsLimit = Math.min(maxLimit, Math.max(recordsLimit + 1, getIncreasedLimit(oldLimit)));
+
+            if (LOGGER.isInfoEnabled()) {
+                LOGGER.info(KeyValueLogMessage.build("Re-increasing limit of online index build",
+                                LogMessageKeys.INDEX_NAME, common.getTargetIndexesNames(),
+                                LogMessageKeys.INDEXER_ID, common.getUuid(),
+                                LogMessageKeys.INDEX_NAME, common.getTargetIndexesNames(),
+                                LogMessageKeys.OLD_LIMIT, oldLimit)
+                        .addKeysAndValues(logMessageKeyValues())
```

Comment:
Should this have an `additionalLogMessageKeyValues` parameter like `decreaseLimit`

Reply:
✅